### ISSUE TITLE
Activation offload fails when mesh rebinding encounters MemorySpace.Host

### ISSUE THREAD

**Issue by dlwh** [NVIDIA/jaxpp#13]
Activation offloading with `jax.checkpoint_policies.save_and_offload_only_these_names` fails during pipeline mesh rebinding. `replace_captured_meshes` passes a `device_put` target of `MemorySpace.Host` to `updated_named_sharding_mesh`, which asserts that the target is a `NamedSharding`. The backward rematerialization also contains a `MemorySpace.Device` transfer.

Reproduced on JAX 0.11.0 with JAXPP `46b8443eed01f54688dd24ae451203a504b1cff8` and current main `78458f8b258f0fd467be13b4b8e7056a853e320b`. The original failure occurred in a distributed DualPipeV run; the following test reproduces it on CPU without a model or distributed initialization.

Reproduce:
1. In a JAXPP environment, save this as `test_offload.py`.
2. Run `JAX_PLATFORMS=cpu python -m pytest test_offload.py -q`.

```python
from functools import partial

import jax
import jax.numpy as jnp
import numpy as np
from jax.ad_checkpoint import checkpoint_name
from jaxpp import jax_compat as jc
from jaxpp.core import replace_captured_meshes


def test_replace_captured_meshes_with_activation_offload():
    policy = jax.checkpoint_policies.save_and_offload_only_these_names(
        names_which_can_be_saved=(),
        names_which_can_be_offloaded=("carry",),
        offload_src="device",
        offload_dst="pinned_host",
    )

    @partial(jax.checkpoint, policy=policy)
    def loss(x):
        return jnp.sin(checkpoint_name(x, "carry")).sum()

    x = jnp.arange(4, dtype=jnp.float32)
    mesh = jax.sharding.Mesh(np.array(jax.devices()[:1]), ("x",))
    closed_jaxpr = jax.make_jaxpr(jax.value_and_grad(loss))(x)
    rebound = replace_captured_meshes(closed_jaxpr, mesh)
    actual_loss, actual_grad = jc.core.eval_jaxpr(rebound.jaxpr, rebound.consts, x)
    np.testing.assert_allclose(actual_loss, jnp.sin(x).sum())
    np.testing.assert_allclose(actual_grad, jnp.cos(x))
```

The failure is:

```text
src/jaxpp/core.py:727: in replace_captured_meshes
    "devices": updated_named_sharding_mesh(eqn.params["devices"], new_mesh)
src/jaxpp/utils.py:237: in update_leaf
    assert isinstance(s, jax.sharding.NamedSharding)
AssertionError  # s = MemorySpace.Host
```

Expected: mesh rebinding preserves memory-space-only transfers and the loss/gradient match the reference.

A local fix in the `device_put_p` branch preserves `jcore.MemorySpace` leaves while continuing to rebind other targets:

```python
"devices": jax.tree.map(
    lambda target: target
    if isinstance(target, jcore.MemorySpace)
    else updated_named_sharding_mesh(target, new_mesh),
    eqn.params["devices"],
)
```

This local patch targets JAX 0.11.0; an upstream implementation also needs to support older JAX versions that do not expose `core.MemorySpace`. The test fails before this change and passes after it on current main. The same change also allowed the original 16-H100 DualPipeV smoke run, with activation and optimizer-state offloading enabled, to finish three finite training updates. Reporting the reproducer here in accordance with CONTRIBUTING.md's policy on external code contributions.

**Comment by Angelogeb** [comment]
Thanks for the detailed report! We'll fix it.

**Comment by Angelogeb** [comment]
This is [now fixed](https://github.com/NVIDIA/jaxpp/commit/328f75a80cecf22c7cc030a82d8941d3c1e220b6). Feel free to follow up if the issue isn't solved, otherwise please close it.
